rust/lance-file/src/reader.rs

```diff
@@ -0,0 +1,1800 @@
+        // we need this m variable here so the else block has a place to store the read
+        // manifest value. If we declare this var in the else block, it won't live long
+        // enough and the returned ref will be invalid.
+        let mut m: Manifest;
+        let manifest = if let Some(m) = manifest {
+            m
+        } else {
+            m = read_struct(object_reader.as_ref(), metadata.manifest_position.unwrap()).await?;
+            m.schema.load_dictionary(object_reader.as_ref()).await?;
+            &m
+        };
+
+        // read the fragment metadata so we can handle deletion
+        let fragment = if is_dataset {
+            let fragment = manifest
+                .fragments
+                .iter()
+                .find(|frag| frag.id == fragment_id)
+                .map(|f| f.to_owned())
+                .ok_or(Error::IO {
+                    message: format!("Fragment {} not found in manifest", fragment_id),
+                    location: location!(),
+                })?;
+            Some(fragment)
+        } else {
+            None
+        };
+
+        let page_table = async {
+            Self::load_from_cache(session, path, |_| async {
+                // TODO: we should have a more efficient way to look up the fields
+                // present in a data file. We might want to include this info in the
+                // data file's metadata.
+                let field_ids = if let Some(fragment) = fragment.as_ref() {
+                    Cow::Borrowed(
+                        &fragment
+                            .files
+                            .iter()
+                            .find(|f| path.to_string().ends_with(&f.path))
+                            .ok_or_else(|| Error::Internal {
+                                message: format!(
+                                    "File {} not found in fragment {:?}",
+                                    path, fragment
+                                ),
+                                location: location!(),
+                            })?
+                            .fields,
+                    )
+                } else {
+                    let max_id = manifest.schema.max_field_id().unwrap() as i32 + 1;
+                    Cow::Owned((0..max_id).collect::<Vec<i32>>())
+                };
+
+                PageTable::load(
+                    object_reader.as_ref(),
+                    metadata.page_table_position,
+                    field_ids.len() as i32,
+                    metadata.num_batches() as i32,
+                    field_ids[0],
+                )
+                .await
+            })
+            .await
+        };
+
+        let projection = manifest.schema.clone();
+
+        let deletion_vector = async {
+            if let Some(fragment) = &fragment {
+                Self::load_deletion_vector(object_store, fragment, session).await
+            } else {
+                Ok(None)
+            }
+        };
```

Comment:
Also confused here because I though the deletion vector was being loaded at the fragment level now.

Reply:
Wait...`lance-file` shouldn't even be in this PR.  I think I messed up a rebase.  Let me take it out :laughing:

Reply:
Ok, sorry.  `lance-file` is now gone.  It will return in #1852